IllegalStateException: Calling invokeAndWait from read-action leads to possible deadlock.

- IntelliJ Ultimate 2019.2.3
- HaskForce 0.3.43

Got a crash from an unhandled exception, in this source file below. 

```
ExternalToolPass: 

com.intellij.diagnostic.PluginException: annotator: com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator@6e6274b7 (class com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator) [Plugin: com.haskforce]
	at com.intellij.ide.plugins.PluginManagerCore.createPluginException(PluginManagerCore.java:403)
	at com.intellij.diagnostic.PluginProblemReporterImpl.createPluginExceptionByClass(PluginProblemReporterImpl.java:12)
	at com.intellij.diagnostic.PluginException.createByClass(PluginException.java:72)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.process(ExternalToolPass.java:253)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.doAnnotate(ExternalToolPass.java:206)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.getInfos(ExternalToolPass.java:157)
	at com.intellij.codeInsight.daemon.impl.DaemonCodeAnalyzerImpl.runMainPasses(DaemonCodeAnalyzerImpl.java:259)
	at com.intellij.openapi.vcs.impl.CodeSmellDetectorImpl.lambda$runMainPasses$5(CodeSmellDetectorImpl.java:180)
	at com.intellij.openapi.project.DumbService.lambda$runReadActionInSmartMode$0(DumbService.java:108)
	at com.intellij.openapi.project.DumbService.lambda$runReadActionInSmartMode$1(DumbService.java:153)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:936)
	at com.intellij.openapi.application.ReadAction.compute(ReadAction.java:57)
	at com.intellij.openapi.project.DumbService.runReadActionInSmartMode(DumbService.java:146)
	at com.intellij.openapi.project.DumbService.runReadActionInSmartMode(DumbService.java:108)
	at com.intellij.openapi.vcs.impl.CodeSmellDetectorImpl.runMainPasses(CodeSmellDetectorImpl.java:180)
	at com.intellij.openapi.vcs.impl.CodeSmellDetectorImpl.lambda$findCodeSmells$4(CodeSmellDetectorImpl.java:161)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:169)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:591)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:537)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:59)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:156)
	at com.intellij.openapi.vcs.impl.CodeSmellDetectorImpl.findCodeSmells(CodeSmellDetectorImpl.java:160)
	at com.intellij.openapi.vcs.impl.CodeSmellDetectorImpl.findCodeSmells(CodeSmellDetectorImpl.java:138)
	at com.intellij.openapi.vcs.impl.CodeSmellDetectorImpl.access$000(CodeSmellDetectorImpl.java:43)
	at com.intellij.openapi.vcs.impl.CodeSmellDetectorImpl$1.run(CodeSmellDetectorImpl.java:100)
	at com.intellij.openapi.progress.impl.CoreProgressManager$TaskRunnable.run(CoreProgressManager.java:894)
	at com.intellij.openapi.progress.impl.CoreProgressManager$5.run(CoreProgressManager.java:447)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:169)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:591)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:537)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:59)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:156)
	at com.intellij.openapi.application.impl.ApplicationImpl.lambda$null$9(ApplicationImpl.java:552)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.run(ApplicationImpl.java:294)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalStateException: Calling invokeAndWait from read-action leads to possible deadlock.
	at com.intellij.openapi.application.impl.ApplicationImpl.invokeAndWait(ApplicationImpl.java:635)
	at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator.saveAllFiles(HaskellExternalAnnotator.scala:58)
	at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator.doAnnotate(HaskellExternalAnnotator.scala:39)
	at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator.doAnnotate(HaskellExternalAnnotator.scala:16)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.doAnnotate(ExternalToolPass.java:203)
	... 34 more
```

```
{-# OPTIONS_GHC -Wall #-}
{-# LANGUAGE OverloadedStrings #-}
module Generate.JavaScript.Name
  ( Name(..) -- only for use by Generate.TypedJavaScript.Name
  , toBuilder
  , fromIndex
  , fromInt
  , fromLocal
  , fromGlobal
  , fromCycle
  , fromKernel
  , makeF
  , makeA
  , makeLabel
  , makeTemp
  , dollar
  )
  where


import qualified Data.ByteString.Builder as B
import Data.Monoid ((<>))
import qualified Data.Map as Map
import qualified Data.Name as Name
import qualified Data.Set as Set
import qualified Data.Utf8 as Utf8
import Data.Word (Word8)

import qualified Data.Index as Index
import qualified Elm.ModuleName as ModuleName
import qualified Elm.Package as Pkg


import Elm

-- NAME


newtype Name =
  Name { toBuilder :: B.Builder }
  deriving (Show)

instance Eq Name where
  (==) (Name a) (Name b) = B.toLazyByteString a == B.toLazyByteString b

-- CONSTRUCTORS


fromIndex :: Index.ZeroBased -> Name
fromIndex index =
  fromInt (Index.toMachine index)


fromInt :: Int -> Name
fromInt n =
  Name (Name.toBuilder (intToAscii n))


fromLocal :: Name.Name -> Name
fromLocal name =
  if Set.member name reservedNames then
    Name ("_" <> Name.toBuilder name)
  else
    Name (Name.toBuilder name)


fromGlobal :: ModuleName.Canonical -> Name.Name -> Name
fromGlobal home name =
  Name $ homeToBuilder home <> usd <> Name.toBuilder name


fromCycle :: ModuleName.Canonical -> Name.Name -> Name
fromCycle home name =
  Name $ homeToBuilder home <> "$cyclic$" <> Name.toBuilder name


fromKernel :: Name.Name -> Name.Name -> Name
fromKernel home name =
  Name ("_" <> Name.toBuilder home <> "_" <> Name.toBuilder name)


{-# INLINE homeToBuilder #-}
homeToBuilder :: ModuleName.Canonical -> B.Builder
homeToBuilder (ModuleName.Canonical (Pkg.Name author project) home) =
  usd <>
  Utf8.toEscapedBuilder 0x2D {- - -} 0x5F {- _ -} author
  <> usd <>
  Utf8.toEscapedBuilder 0x2D {- - -} 0x5F {- _ -} project
  <> usd <>
  Utf8.toEscapedBuilder 0x2E {- . -} 0x24 {- $ -} home



-- TEMPORARY NAMES


makeF :: Int -> Name
makeF n =
  Name ("F" <> B.intDec n)


makeA :: Int -> Name
makeA n =
  Name ("A" <> B.intDec n)


makeLabel :: Name.Name -> Int -> Name
makeLabel name index =
  Name (Name.toBuilder name <> usd <> B.intDec index)


makeTemp :: Name.Name -> Name
makeTemp name =
  Name ("$temp$" <> Name.toBuilder name)


dollar :: Name
dollar =
  Name usd


usd :: B.Builder
usd =
  Name.toBuilder Name.dollar



-- RESERVED NAMES


{-# NOINLINE reservedNames #-}
reservedNames :: Set.Set Name.Name
reservedNames =
  Set.union jsReservedWords elmReservedWords


jsReservedWords :: Set.Set Name.Name
jsReservedWords =
  Set.fromList
    [ "do", "if", "in"
    , "NaN", "int", "for", "new", "try", "var", "let"
    , "null", "true", "eval", "byte", "char", "goto", "long", "case", "else", "this", "void", "with", "enum"
    , "false", "final", "float", "short", "break", "catch", "throw", "while", "class", "const", "super", "yield"
    , "double", "native", "throws", "delete", "return", "switch", "typeof", "export", "import", "public", "static"
    , "boolean", "default", "finally", "extends", "package", "private"
    , "Infinity", "abstract", "volatile", "function", "continue", "debugger", "function"
    , "undefined", "arguments", "transient", "interface", "protected"
    , "instanceof", "implements"
    , "synchronized"
    ]


elmReservedWords :: Set.Set Name.Name
elmReservedWords =
  Set.fromList
    [ "F2", "F3", "F4", "F5", "F6", "F7", "F8", "F9"
    , "A2", "A3", "A4", "A5", "A6", "A7", "A8", "A9"
    ]



-- INT TO ASCII


intToAscii :: Int -> Name.Name
intToAscii n =
  if n < 53 then -- skip $ as a standalone name
    Name.fromWords [toByte n]

  else
    intToAsciiHelp 2 (numStartBytes * numInnerBytes) allBadFields (n - 53)


intToAsciiHelp :: Int -> Int -> [BadFields] -> Int -> Name.Name
intToAsciiHelp width blockSize badFields n =
  case badFields of
    [] ->
      if n < blockSize then
        unsafeIntToAscii width [] n
      else
        intToAsciiHelp (width + 1) (blockSize * numInnerBytes) [] (n - blockSize)

    BadFields renamings : biggerBadFields ->
      let availableSize = blockSize - Map.size renamings in
      if n < availableSize then
        let name = unsafeIntToAscii width [] n in
        Map.findWithDefault name name renamings
      else
        intToAsciiHelp (width + 1) (blockSize * numInnerBytes) biggerBadFields (n - availableSize)



-- UNSAFE INT TO ASCII


unsafeIntToAscii :: Int -> [Word8] -> Int -> Name.Name
unsafeIntToAscii width bytes n =
  if width <= 1 then
    Name.fromWords (toByte n : bytes)
  else
    let
      (quotient, remainder) =
        quotRem n numInnerBytes
    in
    unsafeIntToAscii (width - 1) (toByte remainder : bytes) quotient



-- ASCII BYTES


numStartBytes :: Int
numStartBytes =
  54


numInnerBytes :: Int
numInnerBytes =
  64


toByte :: Int -> Word8
toByte n
  | n < 26  = fromIntegral (97 + n     ) {- lower -}
  | n < 52  = fromIntegral (65 + n - 26) {- upper -}
  | n == 52 = 95 {- _ -}
  | n == 53 = 36 {- $ -}
  | n < 64  = fromIntegral (48 + n - 54) {- digit -}
  | True    = error $ "cannot convert int " ++ show n ++ " to ASCII"



-- BAD FIELDS


newtype BadFields =
  BadFields { _renamings :: Renamings }


type Renamings =
  Map.Map Name.Name Name.Name


allBadFields :: [BadFields]
allBadFields =
  let
    add keyword dict =
      Map.alter (Just . addRenaming keyword) (Utf8.size keyword) dict
  in
    Map.elems $ Set.foldr add Map.empty jsReservedWords


addRenaming :: Name.Name -> Maybe BadFields -> BadFields
addRenaming keyword maybeBadFields =
  let
    width = Utf8.size keyword
    maxName = numStartBytes * numInnerBytes ^ (width - 1) - 1
  in
  case maybeBadFields of
    Nothing ->
      BadFields $ Map.singleton keyword (unsafeIntToAscii width [] maxName)

    Just (BadFields renamings) ->
      BadFields $ Map.insert keyword (unsafeIntToAscii width [] (maxName - Map.size renamings)) renamings
```

--- a/src/com/haskforce/highlighting/annotation/external/GhcModi.java
+++ b/src/com/haskforce/highlighting/annotation/external/GhcModi.java
@@ -23,18 +23,17 @@ import com.intellij.openapi.module.Module;
 import com.intellij.openapi.module.ModuleComponent;
 import com.intellij.openapi.module.ModuleUtilCore;
 import com.intellij.openapi.project.Project;
-import com.intellij.openapi.util.Pair;
+import com.intellij.openapi.util.Computable;
 import com.intellij.openapi.util.text.StringUtil;
+import com.intellij.openapi.vfs.VirtualFile;
 import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
 import com.intellij.xml.util.XmlUtil;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 import scala.Option;
 
 import java.io.*;
-import java.nio.charset.StandardCharsets;
-import java.nio.file.Files;
-import java.nio.file.Paths;
 import java.util.*;
 import java.util.concurrent.*;
 import java.util.regex.Pattern;
@@ -45,7 +44,10 @@ import java.util.regex.Pattern;
 public class GhcModi implements ModuleComponent, SettingsChangeNotifier {
 
     public static Option<GhcModi> get(PsiElement element) {
-        final Module module = ModuleUtilCore.findModuleForPsiElement(element);
+        final Module module =
+            ApplicationManager.getApplication().runReadAction((Computable<Module>) () ->
+                ModuleUtilCore.findModuleForPsiElement(element)
+            );
         if (module == null) return Option.apply(null);
         return get(module);
     }
@@ -112,54 +114,49 @@ public class GhcModi implements ModuleComponent, SettingsChangeNotifier {
         return path != null && !path.trim().isEmpty();
     }
 
-    /**
-     * Checks the module with ghc-modi and returns Problems to be annotated in the source.
-     */
     @Nullable
-    public Future<Problems> check(final @NotNull String file) {
-        return handleGhcModiCall(new GhcModiCallable<Problems>() {
-            @Override
-            public Problems call() throws GhcModiError {
-                return unsafeCheck(file);
-            }
-        });
+    public Problems check(final @NotNull PsiFile psiFile) {
+        return runSync(() -> unsafeCheck(psiFile));
     }
 
     @Nullable
-    public Problems syncCheck(final @NotNull String file) {
-        return runSync(new GhcModiCallable<Problems>() {
-            @Override
-            public Problems call() throws GhcModiError {
-                return unsafeCheck(file);
-            }
-        });
+    private Problems unsafeCheck(final @NotNull PsiFile psiFile) throws GhcModiError {
+        FileSource fileSource = FileSource.fromPsiFile(psiFile);
+        if (fileSource == null) return null;
+        mapFileContents(fileSource);
+        final String stdout = simpleExec("check " + fileSource.path);
+        unmapFileContents(fileSource.path);
+        return stdout == null ? new Problems() : handleCheck(module, fileSource.path, stdout);
     }
 
-    @Nullable
-    private Problems unsafeCheck(final @NotNull String file) throws GhcModiError {
-        // Check the file content so only call out to ghc-mod when the file changes.
-        final String fileContent;
-        try {
-            fileContent = new String(Files.readAllBytes(Paths.get(file)), StandardCharsets.UTF_8);
-        } catch (IOException e) {
-            e.printStackTrace();
-            throw new CheckError(file, e.getMessage());
+    private void mapFileContents(final @NotNull FileSource fileSource) throws GhcModiError {
+        // Terminate with \n then \004 to signal EOF to ghc-modi
+        simpleExec("map-file " + fileSource.path + "\n" + fileSource.contents + "\n\004");
+    }
+
+    private void unmapFileContents(final @NotNull String path) throws GhcModiError {
+        simpleExec("unmap-file " + path);
+    }
+
+    private final static class FileSource {
+        public final @NotNull String path;
+        public final @NotNull String contents;
+
+        private FileSource(@NotNull String path, @NotNull String contents) {
+            this.path = path;
+            this.contents = contents;
         }
 
-        // Check the cache to determine if the file changed.
-        Pair<String, Problems> item = checkCache.get(file);
-        if (item != null && item.first.equals(fileContent)) return item.second;
-
-        // Call out to ghc-mod.
-        final String stdout = simpleExec("check " + file);
-        Problems result = stdout == null ? new Problems() : handleCheck(module, file, stdout);
-
-        // Update the cache.
-        checkCache.put(file, new Pair<>(fileContent, result));
-        return result;
+        static @Nullable FileSource fromPsiFile(PsiFile psiFile) {
+            return ApplicationManager.getApplication().runReadAction((Computable<FileSource>) () -> {
+                VirtualFile vFile = psiFile.getVirtualFile();
+                if (vFile == null) return null;
+                String path = vFile.getCanonicalPath();
+                if (path == null) return null;
+                return new FileSource(path, psiFile.getText());
+            });
+        }
     }
-
-    private ConcurrentHashMap<String, Pair<String, Problems>> checkCache = new ConcurrentHashMap<>();
 
     @Nullable
     public String[] syncLang() {
@@ -509,7 +506,6 @@ public class GhcModi implements ModuleComponent, SettingsChangeNotifier {
         // Clear the caches.
         browseCache.clear();
         findCache.clear();
-        checkCache.clear();
         listCache = null;
 
         // Kill the process, allowing it to be started up again.
@@ -579,6 +575,13 @@ public class GhcModi implements ModuleComponent, SettingsChangeNotifier {
 
     interface GhcModiCallable<V> extends Callable<V> {
         V call() throws GhcModiError;
+    }
+
+    interface GhcModiRunnable extends GhcModiCallable<Void> {
+        void run() throws GhcModiError;
+
+        @Override
+        default Void call() throws GhcModiError { run(); return null; }
     }
 
     static abstract class GhcModiError extends Exception {

--- a/src/com/haskforce/highlighting/annotation/external/HLint.java
+++ b/src/com/haskforce/highlighting/annotation/external/HLint.java
@@ -13,11 +13,13 @@ import com.haskforce.utils.*;
 import com.haskforce.utils.ExecUtil.ExecError;
 import com.intellij.execution.configurations.GeneralCommandLine;
 import com.intellij.execution.configurations.ParametersList;
+import com.intellij.ide.util.PropertiesComponent;
 import com.intellij.lang.annotation.Annotation;
 import com.intellij.notification.NotificationType;
 import com.intellij.openapi.application.ApplicationManager;
 import com.intellij.openapi.diagnostic.Logger;
 import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.Computable;
 import com.intellij.openapi.util.TextRange;
 import com.intellij.openapi.util.text.StringUtil;
 import com.intellij.psi.PsiFile;
@@ -50,16 +52,17 @@ public class HLint {
 
     @NotNull
     public static Problems lint(final @NotNull Project project, @NotNull String workingDirectory,
-                                @NotNull String file, @NotNull HaskellFile haskellFile) {
-        final HaskellToolsConsole toolConsole = HaskellToolsConsole.get(project);
+                                @NotNull HaskellFile haskellFile) {
+        final HaskellToolsConsole.Curried toolConsole = HaskellToolsConsole.get(project).curry(ToolKey.HLINT_KEY);
+        final PropertiesComponent props = PropertiesComponent.getInstance(project);
         final String hlintPath = ToolKey.HLINT_KEY.getPath(project);
         final String hlintFlags = ToolKey.HLINT_KEY.getFlags(project);
         if (hlintPath == null) return new Problems();
 
         return EitherUtil.valueOr(
-            parseProblems(toolConsole, workingDirectory, hlintPath, hlintFlags, file, haskellFile),
+            parseProblems(toolConsole, workingDirectory, hlintPath, hlintFlags, haskellFile),
             e -> {
-                toolConsole.writeError(ToolKey.HLINT_KEY, e.getMessage());
+                toolConsole.writeError(e.getMessage());
                 NotificationUtil.displayToolsNotification(
                   NotificationType.ERROR, project, "hlint", e.getMessage()
                 );
@@ -69,28 +72,30 @@ public class HLint {
     }
 
     @NotNull
-    public static Either<ExecError, Problems> parseProblems(final HaskellToolsConsole toolConsole,
+    public static Either<ExecError, Problems> parseProblems(final HaskellToolsConsole.Curried toolConsole,
                                                             final @NotNull String workingDirectory,
                                                             final @NotNull String path,
                                                             final @NotNull String flags,
-                                                            final @NotNull String file,
                                                             final @NotNull HaskellFile haskellFile) {
         return getVersion(toolConsole, workingDirectory, path).flatMap(version -> {
             final boolean useJson = version.$greater$eq(HLINT_MIN_VERSION_WITH_JSON_SUPPORT);
-            final String[] params = getParams(file, haskellFile, useJson);
-            return runHlint(toolConsole, workingDirectory, path, flags, params).map(stdout -> {
-                toolConsole.writeOutput(ToolKey.HLINT_KEY, stdout);
+            final String[] params = getParams(haskellFile, useJson);
+            final String fileContents =
+              ApplicationManager.getApplication().runReadAction(
+                (Computable<String>) haskellFile::getText
+              );
+            return runHlint(toolConsole, workingDirectory, path, flags, fileContents, params).map(stdout -> {
+                toolConsole.writeOutput(stdout);
                 if (useJson) return parseProblemsJson(toolConsole, stdout);
                 return parseProblemsFallback(stdout);
             });
         });
     }
 
-    private static String[] getParams(@NotNull String file,
-                                      @NotNull HaskellFile haskellFile,
+    private static String[] getParams(@NotNull HaskellFile haskellFile,
                                       boolean useJson) {
         final List<String> result = new ArrayList<>(1);
-        result.add(file);
+        result.add("-"); // Read file from stdin
         if (useJson) result.add("--json");
         result.addAll(getParamsFromCabal(haskellFile));
         return result.toArray(new String[0]);
@@ -107,11 +112,11 @@ public class HLint {
      * Parse problems from the hlint --json output.
      */
     @NotNull
-    public static Problems parseProblemsJson(@NotNull HaskellToolsConsole toolsConsole, @NotNull String stdout) {
+    public static Problems parseProblemsJson(@NotNull HaskellToolsConsole.Curried toolsConsole, @NotNull String stdout) {
         return EitherUtil.valueOr(
             parseProblemsJson(stdout),
             e -> {
-                toolsConsole.writeError(ToolKey.HLINT_KEY, e.getMessage());
+                toolsConsole.writeError(e.getMessage());
                 LOG.error(e);
                 return new Problems();
             }
@@ -186,7 +191,7 @@ public class HLint {
     }
 
     @NotNull
-    private static Either<ExecError, VersionTriple> getVersion(HaskellToolsConsole toolConsole, String workingDirectory, String hlintPath) {
+    private static Either<ExecError, VersionTriple> getVersion(HaskellToolsConsole.Curried toolConsole, String workingDirectory, String hlintPath) {
         return HLintUtil.runHLintNumericVersion(toolConsole, workingDirectory, hlintPath);
     }
 
@@ -194,10 +199,11 @@ public class HLint {
      * Runs hlintProg with parameters if hlintProg can be executed.
      */
     @NotNull
-    public static Either<ExecError, String> runHlint(HaskellToolsConsole toolConsole,
+    public static Either<ExecError, String> runHlint(HaskellToolsConsole.Curried toolConsole,
                                                       @NotNull String workingDirectory,
                                                       @NotNull String hlintProg,
                                                       @NotNull String hlintFlags,
+                                                      @Nullable String fileContents,
                                                       @NotNull String... params) {
         GeneralCommandLine commandLine = new GeneralCommandLine();
         commandLine.setWorkDirectory(workingDirectory);
@@ -208,9 +214,9 @@ public class HLint {
         parametersList.add("--no-exit-code");
         parametersList.addParametersString(hlintFlags);
         parametersList.addAll(params);
-        toolConsole.writeInput(ToolKey.HLINT_KEY, "Using working directory: " + workingDirectory);
-        toolConsole.writeInput(ToolKey.HLINT_KEY, commandLine.getCommandLineString());
-        return ExecUtil.readCommandLine(commandLine);
+        toolConsole.writeInput("Using working directory: " + workingDirectory);
+        toolConsole.writeInput(commandLine.getCommandLineString());
+        return ExecUtil.readCommandLine(commandLine, fileContents);
     }
 
     public static class Problem implements HaskellProblem {

--- a/src/com/haskforce/highlighting/annotation/external/HLintUtil.scala
+++ b/src/com/haskforce/highlighting/annotation/external/HLintUtil.scala
@@ -24,11 +24,11 @@ object HLintUtil {
   }
 
   def runHLintNumericVersion(
-    toolConsole: HaskellToolsConsole,
+    toolConsole: HaskellToolsConsole.Curried,
     workDir: String,
     hlintPath: String
   ): Either[ExecError, VersionTriple] = {
-    HLint.runHlint(toolConsole, workDir, hlintPath, "--numeric-version")
+    HLint.runHlint(toolConsole, workDir, hlintPath, "--numeric-version", null)
       .flatMap(parseVersion)
   }
 }

--- a/src/com/haskforce/highlighting/annotation/external/HaskellExternalAnnotator.scala
+++ b/src/com/haskforce/highlighting/annotation/external/HaskellExternalAnnotator.scala
@@ -1,6 +1,5 @@
 package com.haskforce.highlighting.annotation.external
 
-import com.haskforce.Implicits._
 import com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator.State
 import com.haskforce.highlighting.annotation.{HaskellAnnotationHolder, HaskellProblem, Problems}
 import com.intellij.lang.annotation.{AnnotationHolder, ExternalAnnotator}
@@ -13,7 +12,8 @@ import org.jetbrains.annotations.{NotNull, Nullable}
 import scala.collection.JavaConverters._
 
 /** Single annotator that calls all external tools used for annotations. */
-class HaskellExternalAnnotator extends ExternalAnnotator[PsiFile, State] {
+class HaskellExternalAnnotator
+  extends ExternalAnnotator[HaskellExternalAnnotator.InitialInfo, State] {
 
   /**
    * The default implementation here is to not annotate files that have lexer/parser errors.
@@ -21,46 +21,51 @@ class HaskellExternalAnnotator extends ExternalAnnotator[PsiFile, State] {
    * appearing.  To get around this, we ignore the `hasErrors` argument.
    */
   @Nullable
-  override def collectInformation
-      (@NotNull file: PsiFile, @NotNull editor: Editor, hasErrors: Boolean): PsiFile = {
-    collectInformation(file)
-  }
+  override def collectInformation(
+    @NotNull file: PsiFile,
+    @NotNull editor: Editor,
+    hasErrors: Boolean
+  ): HaskellExternalAnnotator.InitialInfo = collectInformation(file)
 
   /** Simply returns the file so it will get passed to doAnnotate() */
   @NotNull
-  override def collectInformation(@NotNull file: PsiFile): PsiFile = {
-    file
+  override def collectInformation(
+    @NotNull file: PsiFile
+  ): HaskellExternalAnnotator.InitialInfo = {
+    val problemsProviders = HaskellExternalAnnotator.problemsProviderFactories.flatMap(_.get(file))
+    HaskellExternalAnnotator.InitialInfo(
+      file,
+      problemsProviders
+    )
   }
 
   /** Builds the set of annotations to be applied to the source file. */
   @Nullable
-  override def doAnnotate(@NotNull file: PsiFile): State = {
-    // We need to save the files so external processes will see what we see.
-    saveAllFiles()
+  override def doAnnotate(
+    @NotNull info: HaskellExternalAnnotator.InitialInfo
+  ): State = {
     // Constructs our annotation state for a file given our registered providers.
-    State.buildForFile(file)
+    State.create(info)
   }
 
   /** Applies the annotations from the State to the source file. */
-  override def apply
-      (@NotNull file: PsiFile, state: State, @NotNull holder: AnnotationHolder)
-      : Unit = {
+  override def apply(
+    @NotNull file: PsiFile,
+    state: State,
+    @NotNull holder: AnnotationHolder
+  ): Unit = {
     HaskellExternalAnnotator.createAnnotations(
       file, state.getProblems, new HaskellAnnotationHolder(holder)
-    )
-  }
-
-  private def saveAllFiles(): Unit = {
-    ApplicationManager.getApplication.invokeAndWait(
-      () => FileDocumentManager.getInstance.saveAllDocuments(),
-      // We cannot use ModalityState.any(); see issue #288
-      // https://github.com/carymrobbins/intellij-haskforce/issues/288#issuecomment-319835496
-      ModalityState.defaultModalityState()
     )
   }
 }
 
 object HaskellExternalAnnotator {
+
+  final case class InitialInfo(
+    psiFile: PsiFile,
+    problemsProviders: List[ProblemsProvider]
+  )
 
   /** Registered problem providers for our configured tools. */
   val problemsProviderFactories: List[ProblemsProviderFactory] = List(
@@ -69,11 +74,11 @@ object HaskellExternalAnnotator {
   )
 
   /** Helper to annotate problems in our source from external tools. */
-  def createAnnotations
-      (@NotNull file: PsiFile,
-       problems: Stream[HaskellProblem],
-       @NotNull holder: HaskellAnnotationHolder)
-      : Unit = {
+  def createAnnotations(
+    @NotNull file: PsiFile,
+    problems: Stream[HaskellProblem],
+    @NotNull holder: HaskellAnnotationHolder
+  ): Unit = {
     if (!file.isValid) return
     problems.foreach { _.createAnnotations(file, holder) }
   }
@@ -88,9 +93,8 @@ object HaskellExternalAnnotator {
   }
 
   object State {
-    /** Smart constructor which traverses our factories and obtains problem futures. */
-    def buildForFile(file: PsiFile): State = {
-      State(problemsProviderFactories.flatMap(_.get(file).map(_.getProblems)))
+    def create(info: InitialInfo): State = {
+      State(info.problemsProviders.map(_.getProblems))
     }
   }
 }

--- a/src/com/haskforce/highlighting/annotation/external/impl/GhcModiCompileProblemsProvider.scala
+++ b/src/com/haskforce/highlighting/annotation/external/impl/GhcModiCompileProblemsProvider.scala
@@ -1,37 +1,20 @@
 package com.haskforce.highlighting.annotation.external.impl
 
-import java.util.concurrent.Future
-
-import com.intellij.openapi.module.{Module, ModuleUtilCore}
-import com.intellij.openapi.project.Project
-import com.intellij.psi.PsiFile
-
 import com.haskforce.highlighting.annotation.Problems
-import com.haskforce.highlighting.annotation.external.{ProblemsProvider, GhcModi}
-import com.haskforce.utils.WrappedFuture
+import com.haskforce.highlighting.annotation.external.{GhcModi, ProblemsProvider}
+import com.intellij.psi.PsiFile
 
 class GhcModiCompileProblemsProvider private(
   ghcModi: GhcModi,
-  filePath: String
+  psiFile: PsiFile
 ) extends ProblemsProvider {
 
   override def getProblems: Option[Problems] = {
-    Option(ghcModi.syncCheck(filePath))
+    Option(ghcModi.check(psiFile))
   }
 }
 
 object GhcModiCompileProblemsProvider {
-  def create(psiFile: PsiFile): Option[GhcModiCompileProblemsProvider] = for {
-    ghcModi <- GhcModi.get(psiFile)
-    filePath <- Option(psiFile.getVirtualFile.getCanonicalPath)
-  } yield new GhcModiCompileProblemsProvider(ghcModi, filePath)
-}
-
-final class GhcModiFutureProblems(
-  underlying: Future[Problems], project: Project
-) extends WrappedFuture[Option[Problems]] {
-
-  override def get: Option[Problems] = {
-    Option(GhcModi.getFuture(project, underlying))
-  }
+  def create(psiFile: PsiFile): Option[GhcModiCompileProblemsProvider] =
+    GhcModi.get(psiFile).map(new GhcModiCompileProblemsProvider(_, psiFile))
 }

--- a/src/com/haskforce/highlighting/annotation/external/impl/HLintProblemsProvider.scala
+++ b/src/com/haskforce/highlighting/annotation/external/impl/HLintProblemsProvider.scala
@@ -1,39 +1,35 @@
 package com.haskforce.highlighting.annotation.external.impl
 
-import java.util.concurrent.{ExecutionException, Executors, Future}
+import java.util.concurrent.{ExecutionException, Future}
 
-import com.intellij.notification.NotificationType
-import com.intellij.openapi.diagnostic.Logger
-import com.intellij.openapi.project.Project
-import com.intellij.psi.PsiFile
 import com.haskforce.highlighting.annotation.Problems
 import com.haskforce.highlighting.annotation.external.{HLint, ProblemsProvider}
 import com.haskforce.psi.HaskellFile
 import com.haskforce.utils.{CastUtil, ExecUtil, NotificationUtil, WrappedFuture}
+import com.intellij.notification.NotificationType
+import com.intellij.openapi.diagnostic.Logger
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiFile
 
 class HLintProblemsProvider private(
   project: Project,
   workDir: String,
-  filePath: String,
   haskellFile: HaskellFile
 ) extends ProblemsProvider {
 
   override def getProblems: Option[Problems] = {
-    Option(HLint.lint(project, workDir, filePath, haskellFile))
+    Option(HLint.lint(project, workDir, haskellFile))
   }
 }
 
 object HLintProblemsProvider {
   def create(psiFile: PsiFile): Option[HLintProblemsProvider] = for {
     haskellFile <- CastUtil.down[HaskellFile](psiFile)
-    filePath <- Option(psiFile.getVirtualFile.getCanonicalPath)
     workDir = ExecUtil.guessWorkDir(psiFile)
     project = psiFile.getProject
-  } yield new HLintProblemsProvider(project, workDir, filePath, haskellFile)
+  } yield new HLintProblemsProvider(project, workDir, haskellFile)
 
-  val executorService = Executors.newSingleThreadExecutor()
-
-  val LOG = Logger.getInstance(classOf[HLintProblemsProvider])
+  val LOG: Logger = Logger.getInstance(classOf[HLintProblemsProvider])
 }
 
 class HLintFutureProblems(

